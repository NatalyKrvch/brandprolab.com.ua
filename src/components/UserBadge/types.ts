export interface UserBadgeProps {
  userName: string;
  userPhotoUrl: string;
}
