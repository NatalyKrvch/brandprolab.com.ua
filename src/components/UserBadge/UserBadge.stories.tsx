import type { Meta, StoryObj } from '@storybook/react';

import UserBadge from './UserBadge';

const meta: Meta<typeof UserBadge> = {
  title: 'Components/UserBadge',
  component: UserBadge,
  parameters: {
    controls: { expanded: true },
  },
  argTypes: {
    userName: {
      control: 'text',
      description: 'The name of the user to display.',
      table: {
        type: { summary: 'string' },
      },
    },
    userPhotoUrl: {
      control: 'text',
      description: 'The URL of the user’s profile image.',
      table: {
        type: { summary: 'string' },
      },
    },
    userLink: {
      control: 'text',
      description:
        'The external link to the user’s profile. If empty, the badge is not clickable.',
      table: {
        type: { summary: 'string' },
      },
    },
  },
};

export default meta;
type Story = StoryObj<typeof UserBadge>;

export const WithLink: Story = {
  args: {
    userName: 'Jane Doe',
    userPhotoUrl:
      'https://media.vanityfair.com/photos/659d6933b6da4325190e2751/1:1/w_1333,h_1333,c_limit/Lindsay-Lohan.jpg',
    userLink: 'https://example.com/profile/jane',
  },
};

export const WithoutLink: Story = {
  args: {
    userName: 'Anastasia Smith',
    userPhotoUrl:
      'https://media.vanityfair.com/photos/659d6933b6da4325190e2751/1:1/w_1333,h_1333,c_limit/Lindsay-Lohan.jpg',
    userLink: '',
  },
};
