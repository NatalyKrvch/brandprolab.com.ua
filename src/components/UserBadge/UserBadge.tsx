import clsx from 'clsx';
import Image from 'next/image';
import { twMerge } from 'tailwind-merge';

import {
  USER_BADGE_IMAGE_TEST_ID,
  USER_BADGE_NAME_TEST_ID,
  USER_BADGE_TEST_ID,
} from '@/lib/testIDs';

import { ExternalLink } from '../ExternalLink';
import { UserBadgeProps } from './types';

const UserBadge = ({ userName, userPhotoUrl, userLink }: UserBadgeProps) => {
  const Wrapper = userLink ? ExternalLink : 'div';

  return (
    <Wrapper
      {...(userLink ? { href: userLink } : { href: '' })}
      className={twMerge(
        clsx('flex items-center justify-end', 'gap-5 desktop:gap-6'),
      )}
      data-testid={USER_BADGE_TEST_ID}
    >
      <span
        className={twMerge(
          clsx(
            'text-lg font-medium leading-6 text-gray-dark',
            'desktop:text-xl',
          ),
        )}
        data-testid={USER_BADGE_NAME_TEST_ID}
      >
        {userName}
      </span>
      <div className="relative aspect-square w-14 desktop:w-16">
        <Image
          className="rounded-full object-cover"
          src={userPhotoUrl}
          alt={userName}
          fill
          priority
          data-testid={USER_BADGE_IMAGE_TEST_ID}
        />
      </div>
    </Wrapper>
  );
};

export default UserBadge;
