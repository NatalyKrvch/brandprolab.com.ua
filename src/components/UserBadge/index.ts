export * from './types';
export { default as UserBadge } from './UserBadge';
