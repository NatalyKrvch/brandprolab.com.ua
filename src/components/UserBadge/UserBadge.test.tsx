import '@testing-library/jest-dom';

import { render, screen } from '@testing-library/react';

import {
  USER_BADGE_IMAGE_TEST_ID,
  USER_BADGE_NAME_TEST_ID,
  USER_BADGE_TEST_ID,
} from '@/lib/testIDs';

import UserBadge from './UserBadge';

jest.mock('next/image', () => {
  const MockImage = ({
    src,
    alt,
    className,
  }: {
    src: string;
    alt: string;
    className?: string;
  }) => (
    <img
      src={src}
      alt={alt}
      className={className}
      data-testid={USER_BADGE_IMAGE_TEST_ID}
    />
  );
  MockImage.displayName = 'MockImage';
  return MockImage;
});

jest.mock('../ExternalLink', () => {
  return {
    ExternalLink: ({
      href,
      children,
      ...props
    }: {
      href: string;
      children: React.ReactNode;
    }) => (
      <a href={href} {...props} data-testid={USER_BADGE_TEST_ID}>
        {children}
      </a>
    ),
  };
});

describe('UserBadge component', () => {
  const defaultProps = {
    userName: 'John Doe',
    userPhotoUrl: '/test-image.jpg',
    userLink: 'https://example.com',
  };

  it('renders the UserBadge component', () => {
    const { asFragment } = render(<UserBadge {...defaultProps} />);
    const badgeElement = screen.getByTestId(USER_BADGE_TEST_ID);

    expect(badgeElement).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays the correct user name', () => {
    const { asFragment } = render(<UserBadge {...defaultProps} />);
    const nameElement = screen.getByTestId(USER_BADGE_NAME_TEST_ID);

    expect(nameElement).toHaveTextContent('John Doe');
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the user image with correct attributes', () => {
    const { asFragment } = render(<UserBadge {...defaultProps} />);
    const imageElement = screen.getByTestId(USER_BADGE_IMAGE_TEST_ID);

    expect(imageElement).toBeInTheDocument();
    expect(imageElement).toHaveAttribute('src', '/test-image.jpg');
    expect(imageElement).toHaveAttribute('alt', 'John Doe');
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the correct wrapper element when userLink is provided', () => {
    const { asFragment } = render(<UserBadge {...defaultProps} />);
    const wrapperElement = screen.getByTestId(USER_BADGE_TEST_ID);

    expect(wrapperElement.tagName).toBe('A');
    expect(wrapperElement).toHaveAttribute('href', 'https://example.com');
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders a div wrapper when no userLink is provided', () => {
    const { asFragment } = render(
      <UserBadge userName="John Doe" userPhotoUrl="/test-image.jpg" />,
    );
    const wrapperElement = screen.getByTestId(USER_BADGE_TEST_ID);

    expect(wrapperElement.tagName).toBe('DIV');
    expect(asFragment()).toMatchSnapshot();
  });
});
