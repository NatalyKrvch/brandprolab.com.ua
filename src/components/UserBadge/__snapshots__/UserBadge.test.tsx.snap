// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserBadge component displays the correct user name 1`] = `
<DocumentFragment>
  <a
    class="flex items-center justify-end gap-5 desktop:gap-6"
    data-testid="user-badge-test-id"
    href="https://example.com"
  >
    <span
      class="text-lg font-medium leading-6 text-gray-dark desktop:text-xl"
      data-testid="user-badge-name-test-id"
    >
      John Doe
    </span>
    <div
      class="relative aspect-square w-14 desktop:w-16"
    >
      <img
        alt="John Doe"
        class="rounded-full object-cover"
        data-testid="user-badge-image-test-id"
        src="/test-image.jpg"
      />
    </div>
  </a>
</DocumentFragment>
`;

exports[`UserBadge component renders a div wrapper when no userLink is provided 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-end gap-5 desktop:gap-6"
    data-testid="user-badge-test-id"
  >
    <span
      class="text-lg font-medium leading-6 text-gray-dark desktop:text-xl"
      data-testid="user-badge-name-test-id"
    >
      John Doe
    </span>
    <div
      class="relative aspect-square w-14 desktop:w-16"
    >
      <img
        alt="John Doe"
        class="rounded-full object-cover"
        data-testid="user-badge-image-test-id"
        src="/test-image.jpg"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`UserBadge component renders the UserBadge component 1`] = `
<DocumentFragment>
  <a
    class="flex items-center justify-end gap-5 desktop:gap-6"
    data-testid="user-badge-test-id"
    href="https://example.com"
  >
    <span
      class="text-lg font-medium leading-6 text-gray-dark desktop:text-xl"
      data-testid="user-badge-name-test-id"
    >
      John Doe
    </span>
    <div
      class="relative aspect-square w-14 desktop:w-16"
    >
      <img
        alt="John Doe"
        class="rounded-full object-cover"
        data-testid="user-badge-image-test-id"
        src="/test-image.jpg"
      />
    </div>
  </a>
</DocumentFragment>
`;

exports[`UserBadge component renders the correct wrapper element when userLink is provided 1`] = `
<DocumentFragment>
  <a
    class="flex items-center justify-end gap-5 desktop:gap-6"
    data-testid="user-badge-test-id"
    href="https://example.com"
  >
    <span
      class="text-lg font-medium leading-6 text-gray-dark desktop:text-xl"
      data-testid="user-badge-name-test-id"
    >
      John Doe
    </span>
    <div
      class="relative aspect-square w-14 desktop:w-16"
    >
      <img
        alt="John Doe"
        class="rounded-full object-cover"
        data-testid="user-badge-image-test-id"
        src="/test-image.jpg"
      />
    </div>
  </a>
</DocumentFragment>
`;

exports[`UserBadge component renders the user image with correct attributes 1`] = `
<DocumentFragment>
  <a
    class="flex items-center justify-end gap-5 desktop:gap-6"
    data-testid="user-badge-test-id"
    href="https://example.com"
  >
    <span
      class="text-lg font-medium leading-6 text-gray-dark desktop:text-xl"
      data-testid="user-badge-name-test-id"
    >
      John Doe
    </span>
    <div
      class="relative aspect-square w-14 desktop:w-16"
    >
      <img
        alt="John Doe"
        class="rounded-full object-cover"
        data-testid="user-badge-image-test-id"
        src="/test-image.jpg"
      />
    </div>
  </a>
</DocumentFragment>
`;
